### ISSUE TITLE
'__emscripten_thread_init is not a function' in worker due to race condition

### ISSUE THREAD

**Issue by kg** [emscripten-core/emscripten#16132]
Please include the following in your bug report:

**Version of emscripten/emsdk:**
```cd emsdk && ./emsdk install 3.1.1
Resolving SDK version '3.1.1' to 'sdk-releases-upstream-5ee64de9809592480da01372880ea11debd6c740-64bit'
Installing SDK 'sdk-releases-upstream-5ee64de9809592480da01372880ea11debd6c740-64bit'..
```

When using web worker support, there appears to be a race condition where an emscripten worker will receive the 'run' command from the main thread before a previous 'load' command has finished, and it'll fail when processing the command, typically on this line:

https://github.com/emscripten-core/emscripten/blob/4e130e429ca9e52404c0aa4ddfe9982aeef01bb2/src/worker.js#L189

The error will be that '__emscripten_thread_init is not a function', and the root cause is that Module is not actually initialized yet so it's not possible to call any functions. Depending on your configuration you may win or lose the race (I typically lose the race every time in Firefox on my development VM, but win the race pretty often in Chrome on the same VM.)

It seems to me like the worker command handler needs to use a queue to ensure that if it receives a command like 'run' while it's still busy, it won't immediately try to process it. Right now no measures are taken to ensure that the ```.then(``` above that actually initializes Module has run.

**Comment by kg** [comment]
I was able to work around this issue by manually modifying the generated worker.js to save the Promise instance from inside of 'load' and then ```.then``` it inside the 'run' handler, like this:

```js
let loadPromise = null;

// ...

      loadPromise = {{{ EXPORT_NAME }}}(Module);
      loadPromise.then(function (instance) {
        Module = instance;
      });
    } else if (e.data.cmd === 'run') {
      if (!loadPromise)
        throw new Error("The 'load' command must be processed before 'run'");

// ...

      loadPromise.then(function (instance) {
        const Module = instance.Module;
        Module['__performance_now_clock_drift'] = performance.now() - e.data.time;

// ...
```

**Comment by sbc100** [comment]
Oh dear.. that doesn't sounds good.  I wonder why we have never seen this issue in our testing?   Are you using workers directly, or via pthreads API?

My understanding is that (at least with pthreads), we don't send the `run` message to worker until it has first send the `loaded` message to say that it is ready to run:

https://github.com/emscripten-core/emscripten/blob/1adfc097a650f1e8c9b687f5c924c88c82d56bc8/src/library_pthread.js#L222-L229

And the 'loaded' message is not send until the module has been loaded and `initRuntime()` has been called:

https://github.com/emscripten-core/emscripten/blob/1adfc097a650f1e8c9b687f5c924c88c82d56bc8/src/postamble.js#L274-L275

Perhaps you are using some options/settings that make your setup different any of test case?   Can you share you full link command?

**Comment by kg** [comment]
I'll try and dig up the link command. We're using pthreads and the application is somewhat complex, so it's possible there's a bug somewhere else in our code that's causing this.

**Comment by kg** [comment]
From looking into this more, it appears what's happening is that 'load' is being issued twice, and then 'run' can get issued twice. The problem looks like it might be that unusedWorkers is never emptied out, so something can get loaded into the same worker multiple times:

```javascript
  if (!ENVIRONMENT_IS_PTHREAD) {
   var numWorkersToLoad = PThread.unusedWorkers.length;
   PThread.unusedWorkers.forEach(function(w) {
    PThread.loadWasmModuleToWorker(w, function() {
     if (!--numWorkersToLoad) removeRunDependency("wasm-instantiate");
    });
   });
  }
```

In the above, the unused workers aren't removed from the unusedWorkers list after loadWasmModuleToWorker is called, even though that function definitely is not safe to call twice (it just directly sets an event listener instead of adding it, etc.)

Still trying to figure out why *that's* happening, of course...

**Comment by kg** [comment]
Our emcc command lines all look like this, I don't know which one is the link command.

```  /home/kate/Projects/dotnet-runtime-wasm/src/mono/wasm/emsdk/upstream/emscripten/emcc -DHAVE_CONFIG_H -DNO_GLOBALIZATION_SHIM -DSTATIC_ICU=1 -D_THREAD_SAFE -I/home/kate/Projects/dotnet-runtime-wasm/artifacts/obj -I/home/kate/Projects/dotnet-runtime-wasm/src/native -Imono/mini -Imono/mini/../.. -Imono/mini/../../mono/eglib -I/home/kate/Projects/dotnet-runtime-wasm/src/mono/mono/mini/../.. -I/home/kate/Projects/dotnet-runtime-wasm/src/mono/mono/mini/.. -I/home/kate/Projects/dotnet-runtime-wasm/src/mono/mono/mini/../eglib -I/home/kate/Projects/dotnet-runtime-wasm/src/mono/mono/mini/../sgen -I/home/kate/Projects/dotnet-runtime-wasm/src/mono/mono/mini/../eventpipe -Imono/mini/eventpipe -fexceptions -pthread -D_GNU_SOURCE=1 -I/home/kate/.nuget/packages/microsoft.netcore.runtime.icu.transport/7.0.0-preview.2.22074.5/runtimes/browser-wasm-threads/native/include -std=gnu99 -fno-strict-aliasing -fwrapv -Wall -Wunused -Wmissing-declarations -Wpointer-arith -Wno-cast-qual -Wwrite-strings -Wno-switch -Wno-switch-enum -Wno-unused-value -Wno-attributes -Wmissing-prototypes -Wstrict-prototypes -Wnested-externs -Wno-format-zero-length -Wno-unused-function -Qunused-arguments -Wno-tautological-compare -Wno-parentheses-equality -Wno-self-assign -Wno-return-stack-address -Wno-constant-logical-operand -Wno-zero-length-array -Werror=return-type -Werror-implicit-function-declaration -g   -g -fPIC -fvisibility=hidden -MD -MT mono/mini/CMakeFiles/debugger-objects.dir/__/component/debugger.c.o -MF mono/mini/CMakeFiles/debugger-objects.dir/__/component/debugger.c.o.d -o mono/mini/CMakeFiles/debugger-objects.dir/__/component/debugger.c.o   -c /home/kate/Projects/dotnet-runtime-wasm/src/mono/mono/component/debugger.c```

and according to the logs, the cmake invocations look like:
```/home/kate/Projects/dotnet-runtime-wasm/src/mono/wasm/emsdk/upstream/bin/clang -target wasm32-unknown-emscripten -D__EMSCRIPTEN_PTHREADS__=1 -DEMSCRIPTEN -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-cxx-exceptions -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr -D__EMSCRIPTEN_major__=2 -D__EMSCRIPTEN_minor__=0 -D__EMSCRIPTEN_tiny__=23 -D_LIBCPP_ABI_VERSION=2 -Dunix -D__unix -D__unix__ -Werror=implicit-function-declaration -Xclang -iwithsysroot/include/SDL --sysroot=/home/kate/Projects/dotnet-runtime-wasm/src/mono/wasm/emsdk/upstream/emscripten/cache/sysroot -Xclang -iwithsysroot/include/compat -DHAVE_CONFIG_H -DNO_GLOBALIZATION_SHIM -DSTATIC_ICU=1 -D_THREAD_SAFE -I/home/kate/Projects/dotnet-runtime-wasm/artifacts/obj -I/home/kate/Projects/dotnet-runtime-wasm/src/native -Imono/mini -Imono/mini/../.. -Imono/mini/../../mono/eglib -I/home/kate/Projects/dotnet-runtime-wasm/src/mono/mono/mini/../.. -I/home/kate/Projects/dotnet-runtime-wasm/src/mono/mono/mini/.. -I/home/kate/Projects/dotnet-runtime-wasm/src/mono/mono/mini/../eglib -I/home/kate/Projects/dotnet-runtime-wasm/src/mono/mono/mini/../sgen -I/home/kate/Projects/dotnet-runtime-wasm/src/mono/mono/mini/../eventpipe -Imono/mini/eventpipe -fexceptions -pthread -D_GNU_SOURCE=1 -I/home/kate/.nuget/packages/microsoft.netcore.runtime.icu.transport/7.0.0-preview.2.22074.5/runtimes/browser-wasm-threads/native/include -std=gnu99 -fno-strict-aliasing -fwrapv -Wall -Wunused -Wmissing-declarations -Wpointer-arith -Wno-cast-qual -Wwrite-strings -Wno-switch -Wno-switch-enum -Wno-unused-value -Wno-attributes -Wmissing-prototypes -Wstrict-prototypes -Wnested-externs -Wno-format-zero-length -Wno-unused-function -Qunused-arguments -Wno-tautological-compare -Wno-parentheses-equality -Wno-self-assign -Wno-return-stack-address -Wno-constant-logical-operand -Wno-zero-length-array -Werror=return-type -Werror-implicit-function-declaration -g -g -fPIC -fvisibility=hidden -MD -MT mono/mini/CMakeFiles/debugger-objects.dir/__/component/debugger.c.o -MF mono/mini/CMakeFiles/debugger-objects.dir/__/component/debugger.c.o.d -c /home/kate/Projects/dotnet-runtime-wasm/src/mono/mono/component/debugger.c -o mono/mini/CMakeFiles/debugger-objects.dir/__/component/debugger.c.o```

Debugging this more, the double load may have been a bug in firefox's devtools, because I can't get the log messages to appear twice in chrome. What appears to *actually* be happening now (while the promise thing did work, I reverted it to troubleshoot) is that while ```Module.__emscripten_thread_init``` is not a function, ```Module.Module.__emscripten_thread_init``` is. I think this part may be a bug in our code so I'll look into that.

**Comment by kg** [comment]
OK, I believe I've figured everything out here, and none of it (i think?) is a bug in emscripten, just some room for improvements:

1. We were loading our code twice in workers because some of the generated code was treating a worker like a regular page context, https://github.com/kg/runtime/commit/7142965798a3d16d863d59dc79fa7d514b5a5b02 . This is our fault, but I think it would be really good if emscripten generated some basic checks to throw an error when you attempt to load a module twice. It's very easy to do that and hard to tell that you did it.
2. The signature of our generated constructor function was changed at some point so it no longer returned the bare Module object as emscripten expected - this works in every case but, as it turns out, web workers. So once we got to the point of firing a load event and then a run event in the correct order, it was failing due to the nesting. In this case also I don't consider this an emscripten bug, but the dependency on the exact return value being the Module is a little non-obvious and the error you get isn't great.

If y'all are open to a PR that improves error handling for these cases I can try and put one together.
